NullPointerExeception when running quarkus:dev when trying to write dynamic property from git-commit-id-maven-plugin to manifest

### Describe the bug

Hello,

I am trying to add the current git commit id to the MANIFEST.MF of the quarkus build artifact.

 In the past I have used `git-commit-id-maven-plugin` to extract this value into a maven property:

```
<configuration>
	...
	<injectAllReactorProjects>true</injectAllReactorProjects>
	...
	<includeOnlyProperties>
		<includeOnlyProperty>git.commit.id.describe</includeOnlyProperty>
	</includeOnlyProperties>
</configuration>
```

and add the property `git.commit.id.describe` to the jar file using the `maven-jar-plugin` with:

```
<archive>
	<manifestEntries>
		<Implementation-Build>${git.commit.id.describe}</Implementation-Build>
	</manifestEntries>
</archive>
```

Now from what I understand, with Quarkus I can no longer use the `maven-jar-plugin` but have to use the `quarkus-maven-plugin` to set manifest values:

```
<configuration>
	<manifestEntries>
		<Implementation-Build>${git.commit.id.describe}</Implementation-Build>
	</manifestEntries>
</configuration>
```

When running the project with `mvn quarkus-dev` (or in my case form intellij with the quarkus plugin), I get a NullPointerException on parsing properties and the application does not start.

### Expected behavior

When building the project with `mvn clean package -DskipTests=true`, I expect to find in the quarkus-run.jar MANIFEST.MF file

`Implementation-Build: [my.git.commit.id]`

When running the project with `mvn quarkus-dev` (or in my case form intellij with the quarkus plugin), I expect the application to start

### Actual behavior

When building the project with `mvn clean package -DskipTests=true`, I find in the quarkus-run.jar MANIFEST.MF file

`Implementation-Build: [my.git.commit.id]`

When running the project with `mvn quarkus-dev` (or in my case form intellij with the quarkus plugin), I get:

```
[INFO] Error stacktraces are turned on.
[INFO] Scanning for projects...
[INFO] 
[INFO] ---------------------< org.acme:code-with-quarkus >---------------------
[INFO] Building code-with-quarkus 1.0.0-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- quarkus-maven-plugin:2.16.6.Final:dev (default-cli) @ code-with-quarkus ---
[INFO] Invoking org.apache.maven.plugins:maven-resources-plugin:2.6:resources @ code-with-quarkus
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] Copying 2 resources
[INFO] Invoking io.quarkus.platform:quarkus-maven-plugin:2.16.6.Final:generate-code @ code-with-quarkus
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  1.904 s
[INFO] Finished at: 2023-04-21T12:04:38+02:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal io.quarkus.platform:quarkus-maven-plugin:2.16.6.Final:dev (default-cli) on project code-with-quarkus: Quarkus code generation phase has failed: NullPointerException -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal io.quarkus.platform:quarkus-maven-plugin:2.16.6.Final:dev (default-cli) on project code-with-quarkus: Quarkus code generation phase has failed
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute2 (MojoExecutor.java:375)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute (MojoExecutor.java:351)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:215)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:171)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:163)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:294)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:960)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:293)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:196)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:77)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:568)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
Caused by: org.apache.maven.plugin.MojoExecutionException: Quarkus code generation phase has failed
    at io.quarkus.maven.GenerateCodeMojo.generateCode (GenerateCodeMojo.java:85)
    at io.quarkus.maven.GenerateCodeMojo.doExecute (GenerateCodeMojo.java:53)
    at io.quarkus.maven.QuarkusBootstrapMojo.execute (QuarkusBootstrapMojo.java:154)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.twdata.maven.mojoexecutor.MojoExecutor.executeMojo (MojoExecutor.java:120)
    at io.quarkus.maven.DevMojo.executeIfConfigured (DevMojo.java:655)
    at io.quarkus.maven.DevMojo.triggerPrepare (DevMojo.java:602)
    at io.quarkus.maven.DevMojo.triggerCompile (DevMojo.java:623)
    at io.quarkus.maven.DevMojo.handleAutoCompile (DevMojo.java:582)
    at io.quarkus.maven.DevMojo.execute (DevMojo.java:412)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute2 (MojoExecutor.java:370)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute (MojoExecutor.java:351)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:215)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:171)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:163)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:294)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:960)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:293)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:196)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:77)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:568)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
Caused by: java.lang.NullPointerException
    at java.util.concurrent.ConcurrentHashMap.putVal (ConcurrentHashMap.java:1011)
    at java.util.concurrent.ConcurrentHashMap.put (ConcurrentHashMap.java:1006)
    at java.util.Properties.put (Properties.java:1301)
    at io.quarkus.maven.QuarkusBootstrapProvider$QuarkusMavenAppBootstrap.doBootstrap (QuarkusBootstrapProvider.java:193)
    at io.quarkus.maven.QuarkusBootstrapProvider$QuarkusMavenAppBootstrap.bootstrapApplication (QuarkusBootstrapProvider.java:292)
    at io.quarkus.maven.QuarkusBootstrapProvider.bootstrapApplication (QuarkusBootstrapProvider.java:103)
    at io.quarkus.maven.QuarkusBootstrapMojo.bootstrapApplication (QuarkusBootstrapMojo.java:272)
    at io.quarkus.maven.GenerateCodeMojo.generateCode (GenerateCodeMojo.java:70)
    at io.quarkus.maven.GenerateCodeMojo.doExecute (GenerateCodeMojo.java:53)
    at io.quarkus.maven.QuarkusBootstrapMojo.execute (QuarkusBootstrapMojo.java:154)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.twdata.maven.mojoexecutor.MojoExecutor.executeMojo (MojoExecutor.java:120)
    at io.quarkus.maven.DevMojo.executeIfConfigured (DevMojo.java:655)
    at io.quarkus.maven.DevMojo.triggerPrepare (DevMojo.java:602)
    at io.quarkus.maven.DevMojo.triggerCompile (DevMojo.java:623)
    at io.quarkus.maven.DevMojo.handleAutoCompile (DevMojo.java:582)
    at io.quarkus.maven.DevMojo.execute (DevMojo.java:412)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute2 (MojoExecutor.java:370)
    at org.apache.maven.lifecycle.internal.MojoExecutor.doExecute (MojoExecutor.java:351)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:215)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:171)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:163)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:294)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:960)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:293)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:196)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at jdk.internal.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:77)
    at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:568)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
[ERROR] 
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
```

### How to Reproduce?

Steps to reproduce the behavior:

1. Create a project on code.quarkus.io, maven, Java 17, with resteasy reactive (although that might not be necessary.
2. Add the `git-commit-id-maven-plugin` to the pom
```
<!-- Start: Added git-commit-id-plugin -->
<plugin>
	<groupId>io.github.git-commit-id</groupId>
	<artifactId>git-commit-id-maven-plugin</artifactId>
	<version>5.0.0</version>
	<executions>
		<execution>
			<id>get-git-info</id>
			<goals>
				<goal>revision</goal>
			</goals>
		</execution>
	</executions>
	<configuration>
		<dotGitDirectory>${project.basedir}/.git</dotGitDirectory>
		<verbose>false</verbose>
		<generateGitPropertiesFile>true</generateGitPropertiesFile>
		<injectAllReactorProjects>true</injectAllReactorProjects>

		<includeOnlyProperties>
			<includeOnlyProperty>git.commit.id.describe</includeOnlyProperty>
		</includeOnlyProperties>
	</configuration>
</plugin>
<!-- End: Added git-commit-id-plugin -->
```
3. Add this to the `quarkus-maven-plugin`
```
<!-- Start: Adding 'git.commit.id.describe' property from 'git-commit-id-plugin' -->
<configuration>
	<manifestEntries>
		<Implementation-Build>${git.commit.id.describe}</Implementation-Build>
	</manifestEntries>
</configuration>
<!-- End: Adding 'git.commit.id.describe' property from 'git-commit-id-plugin' -->		
```
4. Run project with `mvn quarkus:dev`

### Output of `uname -a` or `ver`

Microsoft Windows [Version 10.0.22621.1555]

### Output of `java -version`

openjdk version "17.0.5" 2022-10-18 OpenJDK Runtime Environment Temurin-17.0.5+8 (build 17.0.5+8) OpenJDK 64-Bit Server VM Temurin-17.0.5+8 (build 17.0.5+8, mixed mode, sharing)

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

2.16.6

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.8.6 (84538c9988a25aec085021c365c560670ad80f63) Maven home: C:\dev\tools\maven\apache-maven-3.8.6 Java version: 17.0.5, vendor: Eclipse Adoptium, runtime: C:\dev\tools\jdk\adoptium\jdk-17.0.5+8 Default locale: en_US, platform encoding: Cp1252 OS name: "windows 11", version: "10.0", arch: "amd64", family: "windows"

### Additional information

- When removing the `${git.commit.id.describe}` from the `manifestEntries` the application works with quarkus:dev, but I do not get the Implementation-Build entry in the manifest.
- When adding `${git.commit.id.describe}` to the maven `<properties>`, the application works with quarkus:dev, but then the value is static and it is not overwritten by the `git-commit-id-maven-plugin` during build.
- When running `mvn compile quarkus:dev`, the application works, but the intellij plugin does not use this. It likely works as the `compile` phase triggers the `git-commit-id-maven-plugin` to set its properties in the reactor.
- The `${git.commit.id.describe}` is set dynamically by the `git-commit-id-maven-plugin`.

It seems as if the `git-commit-id-maven-plugin` is not run when running `quarkus:dev`, or the properties are not copied the reactor. It is as if the `quarkus:dev` goal does not use the rest of maven at all (seeing how the maven-jar-plugin also is not used).

I am stuck on how to make both these scenarios work at the same time. I am happy to solve this problem in another way, with a different plugin or setting in maven.

--- a/devtools/maven/src/main/java/io/quarkus/maven/DevMojo.java
+++ b/devtools/maven/src/main/java/io/quarkus/maven/DevMojo.java
@@ -102,7 +102,6 @@ import io.quarkus.maven.components.MavenVersionEnforcer;
 import io.quarkus.maven.components.QuarkusWorkspaceProvider;
 import io.quarkus.maven.dependency.ArtifactCoords;
 import io.quarkus.maven.dependency.ArtifactKey;
-import io.quarkus.maven.dependency.GACT;
 import io.quarkus.maven.dependency.ResolvedDependency;
 import io.quarkus.paths.PathList;
 import io.quarkus.runtime.LaunchMode;
@@ -117,47 +116,23 @@ import io.smallrye.common.expression.Expression;
 @Mojo(name = "dev", defaultPhase = LifecyclePhase.PREPARE_PACKAGE, requiresDependencyResolution = ResolutionScope.TEST, threadSafe = true)
 public class DevMojo extends AbstractMojo {
 
-    private static final String PROCESS_CLASSES = "process-classes";
+    private static final Set<String> IGNORED_PHASES = Set.of(
+            "pre-clean", "clean", "post-clean");
 
-    /**
-     * running any one of these phases means the compile phase will have been run, if these have
-     * not been run we manually run compile.
-     */
-    private static final List<String> POST_COMPILE_PHASES = List.of(
+    private static final List<String> PRE_DEV_MODE_PHASES = List.of(
+            "validate",
+            "initialize",
+            "generate-sources",
+            "process-sources",
+            "generate-resources",
+            "process-resources",
             "compile",
-            PROCESS_CLASSES,
+            "process-classes",
             "generate-test-sources",
             "process-test-sources",
             "generate-test-resources",
             "process-test-resources",
-            "test-compile",
-            "process-test-classes",
-            "test",
-            "prepare-package",
-            "package",
-            "pre-integration-test",
-            "integration-test",
-            "post-integration-test",
-            "verify",
-            "install",
-            "deploy");
-
-    /**
-     * running any one of these phases means the test-compile phase will have been run, if these have
-     * not been run we manually run test-compile
-     */
-    private static final List<String> POST_TEST_COMPILE_PHASES = List.of(
-            "test-compile",
-            "process-test-classes",
-            "test",
-            "prepare-package",
-            "package",
-            "pre-integration-test",
-            "integration-test",
-            "post-integration-test",
-            "verify",
-            "install",
-            "deploy");
+            "test-compile");
 
     private static final String IO_QUARKUS = "io.quarkus";
     private static final String QUARKUS_GENERATE_CODE_GOAL = "generate-code";
@@ -364,7 +339,7 @@ public class DevMojo extends AbstractMojo {
     @Component
     private ToolchainManager toolchainManager;
 
-    private Map<GACT, Plugin> pluginMap;
+    private Map<ArtifactKey, Plugin> pluginMap;
 
     @Component
     protected QuarkusBootstrapProvider bootstrapProvider;
@@ -458,8 +433,7 @@ public class DevMojo extends AbstractMojo {
                         getLog().info("Changes detected to " + changed + ", restarting dev mode");
                         final DevModeRunner newRunner;
                         try {
-                            triggerCompile(false, false);
-                            triggerCompile(true, false);
+                            handleAutoCompile();
                             newRunner = new DevModeRunner(runner.launcher.getDebugPortOk());
                         } catch (Exception e) {
                             getLog().info("Could not load changed pom.xml file, changes not applied", e);
@@ -537,15 +511,6 @@ public class DevMojo extends AbstractMojo {
     }
 
     private void handleAutoCompile() throws MojoExecutionException {
-        //we check to see if there was a compile (or later) goal before this plugin
-        boolean compileNeeded = true;
-        boolean testCompileNeeded = true;
-        boolean prepareNeeded = true;
-        boolean prepareTestsNeeded = true;
-
-        String jandexGoalPhase = getGoalPhaseOrNull(IO_SMALLRYE, JANDEX_MAVEN_PLUGIN, JANDEX, PROCESS_CLASSES);
-        String legacyJandexGoalPhase = getGoalPhaseOrNull(ORG_JBOSS_JANDEX, JANDEX_MAVEN_PLUGIN, JANDEX, PROCESS_CLASSES);
-        boolean indexClassNeeded = legacyJandexGoalPhase != null || jandexGoalPhase != null;
 
         List<String> goals = session.getGoals();
         // check for default goal(s) if none were specified explicitly,
@@ -553,67 +518,115 @@ public class DevMojo extends AbstractMojo {
         if (goals.isEmpty() && !StringUtils.isEmpty(project.getDefaultGoal())) {
             goals = List.of(StringUtils.split(project.getDefaultGoal()));
         }
+
+        int latestHandledPhaseIndex = -1;
         for (String goal : goals) {
-            if (goal.endsWith("quarkus:generate-code")) {
-                prepareNeeded = false;
-            }
-            if (goal.endsWith("quarkus:generate-code-tests")) {
-                prepareTestsNeeded = false;
-            }
-
-            if (POST_COMPILE_PHASES.contains(goal)) {
-                compileNeeded = false;
-            }
-
-            if (jandexGoalPhase != null
-                    && POST_COMPILE_PHASES.indexOf(goal) >= POST_COMPILE_PHASES.indexOf(jandexGoalPhase)) {
-                indexClassNeeded = false;
-            }
-            if (jandexGoalPhase == null && legacyJandexGoalPhase != null
-                    && POST_COMPILE_PHASES.indexOf(goal) >= POST_COMPILE_PHASES.indexOf(legacyJandexGoalPhase)) {
-                indexClassNeeded = false;
-            }
-
-            if (POST_TEST_COMPILE_PHASES.contains(goal)) {
-                testCompileNeeded = false;
-            }
             if (goal.endsWith("quarkus:dev")) {
                 break;
             }
-        }
-
-        //if the user did not compile we run it for them
-        if (compileNeeded) {
-            triggerCompile(false, prepareNeeded);
-        }
-        if (indexClassNeeded) {
-            initClassIndexes(jandexGoalPhase == null);
-        }
-        if (testCompileNeeded) {
-            try {
-                triggerCompile(true, prepareTestsNeeded);
-            } catch (Throwable t) {
-                getLog().error("Test compile failed, you will need to fix your tests before you can use continuous testing", t);
+            if (goal.indexOf(':') >= 0 || IGNORED_PHASES.contains(goal)) {
+                continue;
+            }
+            var i = PRE_DEV_MODE_PHASES.indexOf(goal);
+            if (i < 0 || i == PRE_DEV_MODE_PHASES.size() - 1) {
+                // all the necessary goals have already been executed
+                return;
+            }
+            if (i > latestHandledPhaseIndex) {
+                latestHandledPhaseIndex = i;
             }
         }
-    }
 
-    private void initToolchain() throws MojoExecutionException {
-        executeIfConfigured(ORG_APACHE_MAVEN_PLUGINS, MAVEN_TOOLCHAINS_PLUGIN, "toolchain", Map.of());
-    }
+        final String quarkusPluginId = getPluginDescriptor().getId();
+        // configured plugin executions by lifecycle phases
+        final Map<String, List<Map.Entry<Plugin, PluginExecution>>> phaseExecutions = new HashMap<>();
+        // goals with prefixes on the command line
+        final Map<String, Plugin> prefixedGoals = new HashMap<>();
+        for (Plugin p : project.getBuildPlugins()) {
+            if (p.getExecutions().isEmpty()) {
+                continue;
+            }
+            for (PluginExecution e : p.getExecutions()) {
+                if (e.getPhase() != null) {
+                    phaseExecutions.computeIfAbsent(e.getPhase(), k -> new ArrayList<>()).add(Map.entry(p, e));
+                } else {
+                    for (String goal : e.getGoals()) {
+                        if (goal.equals(QUARKUS_GENERATE_CODE_GOAL) && p.getId().equals(quarkusPluginId)) {
+                            var clone = e.clone();
+                            clone.setGoals(List.of(QUARKUS_GENERATE_CODE_GOAL));
+                            // this will schedule it before the compile phase goals
+                            phaseExecutions.computeIfAbsent("compile", k -> new ArrayList<>())
+                                    .add(0, Map.entry(p, clone));
+                        } else if (goal.equals(QUARKUS_GENERATE_CODE_TESTS_GOAL) && p.getId().equals(quarkusPluginId)) {
+                            var clone = e.clone();
+                            clone.setGoals(List.of(QUARKUS_GENERATE_CODE_TESTS_GOAL));
+                            // this will schedule it before the test-compile phase goals
+                            phaseExecutions.computeIfAbsent("test-compile", k -> new ArrayList<>())
+                                    .add(0, Map.entry(p, clone));
+                        } else {
+                            var mojoDescr = getMojoDescriptor(p, goal);
+                            if (mojoDescr.getPhase() != null) {
+                                phaseExecutions.computeIfAbsent(mojoDescr.getPhase(), k -> new ArrayList<>())
+                                        .add(Map.entry(p, e));
+                            } else {
+                                getLog().warn("Failed to determine the lifecycle phase for " + p.getId() + ":" + goal);
+                            }
+                        }
+                    }
+                }
+                if (!e.getGoals().isEmpty()) {
+                    var goalPrefix = getMojoDescriptor(p, e.getGoals().get(0)).getPluginDescriptor().getGoalPrefix();
+                    for (String goal : e.getGoals()) {
+                        prefixedGoals.put(goalPrefix + ":" + goal, p);
+                    }
+                }
+            }
+        }
 
-    private void triggerPrepare(boolean test) throws MojoExecutionException {
-        final PluginDescriptor pluginDescr = getPluginDescriptor();
-        executeIfConfigured(pluginDescr.getGroupId(), pluginDescr.getArtifactId(),
-                test ? QUARKUS_GENERATE_CODE_TESTS_GOAL : QUARKUS_GENERATE_CODE_GOAL,
-                Map.of("mode", LaunchMode.DEVELOPMENT.name(), QuarkusBootstrapMojo.CLOSE_BOOTSTRAPPED_APP, "false"));
-    }
+        // Map<pluginId, List<goals>>
+        final Map<String, List<String>> executedPluginGoals = new HashMap<>();
+        for (String goal : goals) {
+            if (goal.endsWith("quarkus:dev")) {
+                break;
+            }
+            var colon = goal.indexOf(':');
+            if (colon >= 0) {
+                var plugin = prefixedGoals.get(goal);
+                if (plugin == null) {
+                    getLog().warn("Failed to locate plugin for " + goal);
+                } else {
+                    executedPluginGoals.computeIfAbsent(plugin.getId(), k -> new ArrayList<>()).add(goal.substring(colon + 1));
+                }
+            }
+        }
 
-    private void initClassIndexes(boolean legacyJandex) throws MojoExecutionException {
-        if (legacyJandex) {
-            executeIfConfigured(ORG_JBOSS_JANDEX, JANDEX_MAVEN_PLUGIN, JANDEX, Map.of());
-        } else {
-            executeIfConfigured(IO_SMALLRYE, JANDEX_MAVEN_PLUGIN, JANDEX, Map.of());
+        for (int phaseIndex = latestHandledPhaseIndex + 1; phaseIndex < PRE_DEV_MODE_PHASES.size(); ++phaseIndex) {
+            var executions = phaseExecutions.get(PRE_DEV_MODE_PHASES.get(phaseIndex));
+            if (executions == null) {
+                continue;
+            }
+            for (Map.Entry<Plugin, PluginExecution> pe : executions) {
+                var executedGoals = executedPluginGoals.getOrDefault(pe.getKey().getId(), List.of());
+                for (String goal : pe.getValue().getGoals()) {
+                    if (!executedGoals.contains(goal)) {
+                        try {
+                            executeGoal(pe.getKey(), pe.getValue().getId(), goal,
+                                    pe.getKey().getId().equals(quarkusPluginId)
+                                            ? Map.of("mode", LaunchMode.DEVELOPMENT.name(),
+                                                    QuarkusBootstrapMojo.CLOSE_BOOTSTRAPPED_APP, "false")
+                                            : Map.of());
+                        } catch (Throwable t) {
+                            if (goal.equals("testCompile")) {
+                                getLog().error(
+                                        "Test compile failed, you will need to fix your tests before you can use continuous testing",
+                                        t);
+                            } else {
+                                throw t;
+                            }
+                        }
+                    }
+                }
+            }
         }
     }
 
@@ -621,43 +634,27 @@ public class DevMojo extends AbstractMojo {
         return mojoExecution.getMojoDescriptor().getPluginDescriptor();
     }
 
-    private void triggerCompile(boolean test, boolean prepareNeeded) throws MojoExecutionException {
-        handleResources(test);
-
-        if (prepareNeeded) {
-            triggerPrepare(test);
-        }
-
-        // compile the Kotlin sources if needed
-        executeIfConfigured(ORG_JETBRAINS_KOTLIN, KOTLIN_MAVEN_PLUGIN, test ? "test-compile" : "compile", Map.of());
-
-        // Compile the Java sources if needed
-        executeIfConfigured(ORG_APACHE_MAVEN_PLUGINS, MAVEN_COMPILER_PLUGIN, test ? "testCompile" : "compile", Map.of());
-    }
-
-    /**
-     * Execute the resources:resources goal if resources have been configured on the project
-     */
-    private void handleResources(boolean test) throws MojoExecutionException {
-        List<Resource> resources = project.getResources();
-        if (resources.isEmpty()) {
+    private void initToolchain() throws MojoExecutionException {
+        final Plugin plugin = getConfiguredPluginOrNull(ORG_APACHE_MAVEN_PLUGINS, MAVEN_TOOLCHAINS_PLUGIN);
+        if (!isGoalConfigured(plugin, "toolchain")) {
             return;
         }
-        executeIfConfigured(ORG_APACHE_MAVEN_PLUGINS, MAVEN_RESOURCES_PLUGIN, test ? "testResources" : "resources", Map.of());
+        executeGoal(plugin, null, "toolchain", Map.of());
     }
 
-    private void executeIfConfigured(String pluginGroupId, String pluginArtifactId, String goal, Map<String, String> params)
+    private void executeGoal(Plugin plugin, String executionId, String goal, Map<String, String> params)
             throws MojoExecutionException {
-        final Plugin plugin = getConfiguredPluginOrNull(pluginGroupId, pluginArtifactId);
-        if (!isGoalConfigured(plugin, goal)) {
-            return;
+        var msg = new StringBuilder();
+        msg.append("Invoking ").append(plugin.getId()).append(":").append(goal);
+        if (executionId != null) {
+            msg.append(" (").append(executionId).append(")");
         }
-        getLog().info("Invoking " + plugin.getGroupId() + ":" + plugin.getArtifactId() + ":" + plugin.getVersion() + ":" + goal
-                + " @ " + project.getArtifactId());
+        msg.append(" @ ").append(project.getArtifactId());
+        getLog().info(msg.toString());
         executeMojo(
                 plugin(
-                        groupId(pluginGroupId),
-                        artifactId(pluginArtifactId),
+                        groupId(plugin.getGroupId()),
+                        artifactId(plugin.getArtifactId()),
                         version(plugin.getVersion()),
                         plugin.getDependencies()),
                 goal(goal),
@@ -668,22 +665,7 @@ public class DevMojo extends AbstractMojo {
                         pluginManager));
     }
 
-    private String getGoalPhaseOrNull(String groupId, String artifactId, String goal, String defaultPhase) {
-        Plugin plugin = getConfiguredPluginOrNull(groupId, artifactId);
-
-        if (plugin == null) {
-            return null;
-        }
-        for (PluginExecution pluginExecution : plugin.getExecutions()) {
-            if (pluginExecution.getGoals().contains(goal)) {
-                var configuredPhase = pluginExecution.getPhase();
-                return configuredPhase != null ? configuredPhase : defaultPhase;
-            }
-        }
-        return null;
-    }
-
-    public boolean isGoalConfigured(Plugin plugin, String goal) {
+    private boolean isGoalConfigured(Plugin plugin, String goal) {
         if (plugin == null) {
             return false;
         }
@@ -751,10 +733,10 @@ public class DevMojo extends AbstractMojo {
             pluginMap = new HashMap<>();
             // the original plugin keys may include property expressions, so we can't rely on the exact groupId:artifactId keys
             for (Plugin p : project.getBuildPlugins()) {
-                pluginMap.put(new GACT(p.getGroupId(), p.getArtifactId()), p);
+                pluginMap.put(ArtifactKey.ga(p.getGroupId(), p.getArtifactId()), p);
             }
         }
-        return pluginMap.get(new GACT(groupId, artifactId));
+        return pluginMap.get(ArtifactKey.ga(groupId, artifactId));
     }
 
     private Map<Path, Long> readPomFileTimestamps(DevModeRunner runner) throws IOException {

--- a/independent-projects/bootstrap/maven-resolver/src/main/java/io/quarkus/bootstrap/resolver/maven/workspace/Playground.java
+++ b/independent-projects/bootstrap/maven-resolver/src/main/java/io/quarkus/bootstrap/resolver/maven/workspace/Playground.java
@@ -1,0 +1,19 @@
+package io.quarkus.bootstrap.resolver.maven.workspace;
+
+import java.io.File;
+
+import org.eclipse.aether.artifact.DefaultArtifact;
+
+import io.quarkus.bootstrap.resolver.maven.MavenArtifactResolver;
+
+public class Playground {
+
+    public static void main(String[] args) throws Exception {
+
+        MavenArtifactResolver.builder()
+                .setWorkspaceDiscovery(false)
+                .setUserSettings(new File("/home/aloubyansky/playground/quarkus3-proxy-auth-issue/settings.xml"))
+                .build()
+                .resolve(new DefaultArtifact("org.mvnpm", "lit", "pom", "2.7.4"));
+    }
+}

